feat: add support for linux-musl-arm64

aqrln opened the pull request:
Add support for two new binary targets:

- `linux-musl-arm64-openssl-1.1.x`
- `linux-musl-arm64-openssl-3.0.x`

Checklist:

- [x] Engines build images (https://github.com/prisma/engine-images/pull/66)
   - [x] ready
   - [x] merged
- [x] Add the new target in engineer (https://github.com/prisma/engineer/pull/87)
   - [x] ready
   - [x] merged
- [x] Update engineer in prisma-engines to a tagged version once the engineer PR lands, use the branch for testing meanwhile (https://github.com/prisma/prisma-engines/pull/3656)
   - [x] tested with integration branch
   - [x] release engineer 1.47
   - [x] update prisma-engines to the new engineer
- [x] Add support for the new binary targets on the client (this PR)
- [x] Ecosystem tests (https://github.com/prisma/ecosystem-tests/pull/3418)

Closes https://github.com/prisma/prisma/issues/8478

aqrln committed:
feat: add linux-musl-arm64-* to supported platforms

aqrln commented on packages/get-platform/src/getPlatform.ts:
```diff
@@ -74,15 +74,6 @@ export async function getos(): Promise<GetOSResult> {
   const distroInfo = await resolveDistro()
   const archFromUname = await getArchFromUname()
 
-  // TODO: add 'arm64' to the `[...].includes(arch)` check once we have arm64 engines for Alpine
```
I removed this altogether instead of doing what TODO says because we don't have this check for other platforms, and I don't think it's generally useful to hard error on unsupported CPU architecture because users can totally build from source for 32-bit ARM or RISC-V or whatever they need.

jkomyno replied:
Note: we no longer "hard error", we warn now

aqrln replied:
Nice, I missed that, thanks for pointing this out!

aqrln replied:
Should we still have the check for the CPU architecture and just remove the distro from it then?

jkomyno replied:
Let's go with `if (platform === 'linux' && !['x64', 'arm64'].includes(arch)) {`

Drop the `alpine:` prefix from the `alpine:unsupported-arch` warning and change the first sentence to `Prisma only officially supports Linux on `amd64` (x86_64) and `arm64` (aarch64) system architectures`, please.

aqrln replied:
`platform === 'linux'` check was not necessary here since the whole code path here is only relevant for Linux, other than that, done

jkomyno commented:
Added ecosystem-tests here: https://github.com/prisma/ecosystem-tests/pull/3418

aqrln committed:
fix: use FIXED_ENGINES_HASH in download test

This is the only place except one other test script where we don't
pass the specific `version` to `download`, so it defaults to `latest`.

When `latest` is passed, `download` attempts to find the last built
engines version by checking each commit to see if it has all expected
engine files. It does have a special logic for integration branches,
but it requires the corresponding branch on the engine side to be
named without the `integration/` suffix, which is not how we normally
do integration branches nowadays. As a result, it doesn't find the
corresponding engine branch and defaults to `main`, and none of the
commits on `main` have all expected engine files now that we added
`linux-musl-arm64-*` to the list of platforms, so it fails to find any
commit to download the engines for.

jkomyno commented:
Note: I will approve this PR once https://github.com/prisma/engine-images/pull/66 and https://github.com/prisma/engineer/pull/87 are merged

aqrln commented:
Tests fail with last integration version (panic in WASM `get_dmmf` in `generator.test.ts`), presumably because the client side now already includes https://github.com/prisma/prisma/commit/8ad76c3e04bd3d037e2258b6219d062d863ee00c after the rebase, however the engine branch didn't include https://github.com/prisma/prisma-engines/commit/323c8a1169351067817dd9f83076b2f8afad856c yet. Since the PR is already merged and the binaries from `main` should be released soon-ish, those should hopefully work.

aqrln commented:
> those should hopefully work

or rather those should only fail other expected tests since the engine updates are currently blocked on https://github.com/prisma/prisma/pull/17705 / https://github.com/prisma/prisma-engines/pull/3666

jkomyno commented:
I see failures in `views` functional tests, perhaps that's due to the engines being slightly misaligned with the `prisma-engines`' `main`. After all, views are a preview feature and subject to frequent changes these days.
However, we should investigate and confirm/debunk this.

aqrln commented:
@jkomyno it’s expected, see the comment above and https://github.com/prisma/prisma/pull/17705

jkomyno commented:
https://github.com/prisma/prisma/pull/17728 is merged now

aqrln committed:
test: update getPlatform tests

aqrln committed:
test: clarify comment
